app/src/main/java/org/schabi/newpipe/util/potoken/PoTokenWebView.kt

```diff
@@ -1,33 +1,44 @@
 package org.schabi.newpipe.util.potoken
 
 import android.content.Context
-import android.os.Handler
-import android.os.Looper
 import android.util.Log
 import android.webkit.ConsoleMessage
 import android.webkit.JavascriptInterface
 import android.webkit.WebChromeClient
 import android.webkit.WebView
 import androidx.annotation.MainThread
+import androidx.collection.ArrayMap
 import androidx.webkit.WebSettingsCompat
 import androidx.webkit.WebViewFeature
-import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
-import io.reactivex.rxjava3.core.Single
 import io.reactivex.rxjava3.core.SingleEmitter
-import io.reactivex.rxjava3.disposables.CompositeDisposable
-import io.reactivex.rxjava3.schedulers.Schedulers
+import kotlinx.coroutines.CoroutineExceptionHandler
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.MainScope
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.suspendCancellableCoroutine
+import kotlinx.coroutines.withContext
 import org.schabi.newpipe.BuildConfig
 import org.schabi.newpipe.DownloaderImpl
 import java.time.Instant
+import java.time.temporal.ChronoUnit
+import java.util.Collections
+import kotlin.coroutines.Continuation
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
 
 class PoTokenWebView private constructor(
     context: Context,
     // to be used exactly once only during initialization!
-    private val generatorEmitter: SingleEmitter<PoTokenGenerator>,
+    private val continuation: Continuation<PoTokenGenerator>,
 ) : PoTokenGenerator {
     private val webView = WebView(context)
-    private val disposables = CompositeDisposable() // used only during initialization
-    private val poTokenEmitters = mutableListOf<Pair<String, SingleEmitter<String>>>()
+    private val scope = MainScope()
+    private val poTokenContinuations =
+        Collections.synchronizedMap(ArrayMap<String, Continuation<String>>())
```

Comment:
Why do you need a synchronizedMap for this? If the list is only accessed in suspended functions there should be nothing parallel going on, and the ordering of suspend calls should take care of it (as long as everything is run in the same coroutine block)